### DISCUSSION TITLE
WiFi TX power, country max_tx_power and brownouts

### DISCUSSION THREAD

**Discussion by frankonski** [HomeSpan/HomeSpan#613] [Q&A] [answered]
I have been curious (and annoyed) about the brownouts I keep having with an ESP32 WROVER dev board. Even if I am still using a dev board, with breadboard, powered from a 12V alarm control panel, I would not expect brownouts to happen so often (I am yet to see the board running for a week without one).

I have noticed that brownouts seem to happen with wifi events. I learned that wifi can draw a lot of power, so I focused my research out this, and I found something odd that would like to get some eyes on from the HomeSpan dev team (even if this is a team of 1!).

Question to get an answer for: if a country is limited to a `max_tx_power` of `30`, while my ESP32 shows that it can do a `max_tx_power` of `78` (great!!), if the country should be limited to `30`, then why is the board doing `78` at runtime?

I'm using the following code in the main `loop()`:
```
  wifi_country_t country;
  esp_err_t esp_err = ESP_OK;
  static bool country_logged = false;
  int test = CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER;
  int test2 = CONFIG_ESP32_PHY_MAX_TX_POWER;
  static String country_code = "";
  static String prev_country_code = "";
  int8_t max_tx_power = CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER;
  if(!country_logged) {
    if ((esp_err = esp_wifi_get_country(&country)) == ESP_OK) {
      country_code.clear();
      country_code.concat(country.cc[0]);
      country_code.concat(country.cc[1]);
      if (!country_code.equals(prev_country_code)) {
        // Power 78 = 19.5 dBm = 96 milliwatts
        // When country is US, maxTx is set to 30. Does that match to 30dBm?
        // https://www.digitalairwireless.com/articles/blog/wifi-transmit-power-calculations-made-simples
        // So if CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER = 20, what unit is it?
        ESP_ERROR_CHECK(esp_wifi_get_max_tx_power(&max_tx_power));
        WEBLOG("Country code: %s, country maxTxPower: %d (ESP 32 maxTxPower: %d, currentTxPower: %d), policy: %s", 
          country_code, 
          country.max_tx_power,
          max_tx_power,
          WiFi.getTxPower(),
          country.policy == WIFI_COUNTRY_POLICY_AUTO ? "AUTO" : 
            (country.policy == WIFI_COUNTRY_POLICY_MANUAL ? "MANUAL" : "Unknwon"));
        prev_country_code = country_code;
        
      }
    } else {
      country_logged = true;
      WEBLOG("Error getting country: %s", 
      esp_err == ESP_ERR_WIFI_NOT_INIT ? "Wifi not initialized by esp_wifi_init" : "Invalid arg");
    }
  }  
```
Also:
- Why is `CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER` defined to `20` ?

I get the following logged first **prior to connecting to wifi:**
`Country code: CN, country maxTxPower: 20 (ESP 32 maxTxPower: 78, currentTxPower: 78), policy: AUTO`
Then this gets logged **after connection to wifi was successful**
`Country code: US, country maxTxPower: 30 (ESP 32 maxTxPower: 78, currentTxPower: 78), policy: AUTO`

I saw in ESP32 doc that the country can be saved in NVS. 
- Should HomeSpan allow setting the country, so it gets saved between reboots?
- Should HomeSpan set the WiFi max TX Power to the country's max_tx_power?

And could this high `Wifi.getTxPower()` value be the cause of brownouts ? I suspect it...

**Comment by HomeSpan** [comment]
I have not explored WiFi power so would need to do some research to address.  Are you sure you are getting brownouts (as opposed to a reboot due to a software issue, or short that causes a reboot/brownout)?  Though WiFi does take up power, this is usually a problem when running on a battery.  What is the current draw for your power supply?  Also, what board are you using, and how are you powering?

**Reply by frankonski** [reply]
`Are you sure you are getting brownouts`
Until proven otherwise, yes. Since you accepted PR #566, I have been keeping an eye on the _Reset Reason_, and there are just WAY too many brownouts to my liking for what I think is a really simple night time project.

`due to a software issue, or short that causes a reboot/brownout`
I am hunting that darn cause ;-) I believe in hardware that stays always on.

`Though WiFi does take up power, this is usually a problem when running on a battery.`
The alarm panel is connected to a power source, backed by a battery. Could that panel produces power that fluctuates enough to a point where the ESP32 board does not even get 5V? .. because I have a 5v regulator between the panel and the 5V pin of the ESP32.

`What is the current draw for your power supply?`
Which one? where?

`Also, what board are you using`
Freenove ESP32 WROVER DEV.

`how are you powering`
120V --> Alarm panel (+12V aux) --> 5v regulator --> ESP32 5V PIN
 ... and looking at the schematic, the 5V input goes through a 3.3V regulator, to the chip.

The dev board has 2 built-in LED, I have 1 standard LED and one RGB LED on the breadboard. That is pretty much it.

I don't like this `currentTxPower: 78` ....

**Reply by HomeSpan** [reply]
To debug, can you power the ESP32 directly through the serial port with standard wall adapter.  It could be the 5V regulator you are using - what is its capacity?

**Reply by frankonski** [reply]
`can you power the ESP32 directly through the serial port with standard wall adapter.`
Do you power from USB & Apple iPhone wall adapter would do?

`what is its capacity?`
No idea. I don't have electrical enginering skills. I went to a place where they sell this stuff, and they sold me one. :-|

**Reply by HomeSpan** [reply]
USB with Apple Wall Adapter should be fine.

**Reply by frankonski** [reply]
doing it now. Will monitor.

**Reply by frankonski** [reply]
So far so good. The board survived my router's weekly reboot at night, something that the board always ended up browning out on.
Here is something new though. During this router reboot, of course the board loses wifi. When it does, `esp_wifi_get_country()` is back on returning `CN` (China), with country_max_tx_power to 20.
As soon a wifi reconnects, `esp_wifi_get_country()` is back at retuning `US` with country_max_tx_power to 30.
`maxTxPower` and `currentTxPower` remains at 52.

Is there a place in HomeSpan code where, right after AP connect, we could try setting the maxTxPower to what is received by the AP (30 in my case) and set the country (using [esp_wifi_set_country_code](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_wifi.html#_CPPv425esp_wifi_set_country_codePKcb) ), so these values survives reboots?

**Reply by HomeSpan** [reply]
Glad it's working - my guess is the the 5V regulator was not up to the task.   Re the country codes and WiFi power, it may be something that's more hard-coded into the IDF - I have not looked into this portion of the WiFi interface.  Have you tried changing the WiFi power in your sketch?  HomeSpan allows you to specify a callback after WiFi is connected, so you can try setting the power right after the connection is made.

**Reply by frankonski** [reply]
;-) I found that wifi callback functionality before you mentioned it. What I ended up doing is using `set_country_code` and `set_max_tx_power` in that callback. I can confirm that the country code survives reboots, but not the max Tx Power. On reboots it is still way up to 78, so as `WiFi.getTxPower()`.

Let's make an assumption here: lets assume that ESP32 **never** abides to the country's `max_tx_power` value (either the burned in China default value of 20, or the one mapped to the country saved in NVS when setting the country code), this means the ESP32 always go full blast on power on, **except** when the reset code is brownout, which ESP32 lowers the power down to 52 as I saw it earlier. This would explain why so many people on the web seems to disable the brownout detection when wifi starts.

And even at 52, it is still higher than the country's max value.

Quick note, even with a power of 28, so far, I still have have a -48 dBm wifi signal. No change, good enough for my needs.

**Reply by HomeSpan** [reply]
Thanks for the detailed update.   Looks like you have a workable solution - maybe at some point Espressif will address the issue in the IDF and have the country-code determine the max WiFi power at start-up.

**Reply by frankonski** [reply]
For now, I may have nailed it down. I need to read more on this "PHY init data" stuff so power **before** wifi is not too strong. 
While I do my research, I will disconnect the board from Apple USB power adapter and see if I get any more brownouts now that max_tx_power is set to 28 _at runtime_, see how long it will last.

**Reply by frankonski** [reply]
ok, so, I think some magic is happening with the file `sdkconfig`, which I am yet to figure out the best way to control it. BTW, the ESP-IDF VSCode plugin has a really nice UI that explains all settings. But my project builds and _OTA uploads_ without that plugin, so I need to figure out where to set new values in this `sdkconfig` file. Nevertheless.

`esp_wifi_get_max_tx_power()` lowering from 78 to 52 on brownouts is explained by `sdkconfig` having [CONFIG_ESP32_REDUCE_PHY_TX_POWER](https://docs.espressif.com/projects/esp-idf/en/v4.2.4/esp32/api-reference/kconfig.html#config-esp32-reduce-phy-tx-power) set to `y` ... looks like it.
But still, that same `sdkconfig` file has
```
CONFIG_ESP_PHY_MAX_WIFI_TX_POWER=20
CONFIG_ESP_PHY_MAX_TX_POWER=20
```
and power is at 78 on startup after flashing. The PHY maxTxPower should be burned into the application code, as per ESP-IDF doc: "If this option is not enabled (CONFIG_ESP_PHY_INIT_DATA_IN_PARTITION), PHY init data will be embedded into the application binary."

**Reply by HomeSpan** [reply]
Note that the Arduino-ESP core is precompiled by Espressif.  I don't believe any changes made in skconfig, or any changes to any predefined IDF macros will have any effect on a sketch compiled within the Arduino IDE.  You would need to use the IDF tool chain (which by itself will not work for HomeSpan unless you also include the Arduino-ESP32 library).

**Reply by frankonski** [reply]
Knowing that I could lower down the txPower (and keep it down), I disconnected the USB power. It took few minutes (if not seconds) before brownout.
Ok, 5V regulator (7805CV) could now be the culprit, but I should check if I really get 12V from the panel. If I do, then that regulator is probably the culprit.

.. and side note, I am trying to stick to using VSCode rather than Arduino IDE. Just more mature.

**Reply by frankonski** [reply]
I think I understand. Some of Espressif's constants and struct values have a unit of **1 dBm**, while others have a unit of **0.25 dBm**, meaning for some operations, you have to multiply (or divide) by 4.

And the [wifi_power_t](https://github.com/espressif/arduino-esp32/blob/21b88659b9ded3fcc1082f23f498bc2a04cd4f1b/libraries/WiFi/src/WiFiGeneric.h#L111) enum type has _names_ with a 1 dBm unit, while the enum value is 0.25 dBm.

|Function or variable|Unit|Source|
|-------|-------|-------|
|Constant [CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER](https://docs.espressif.com/projects/esp-idf/en/v4.2.4/esp32/api-reference/kconfig.html#config-esp32-phy-max-wifi-tx-power) = 20|1 dBm| |
|struct `wifi_country_t`.`max_tx_power` = 20 by default at initial power on|1 dBm|[wifi_count_t](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_wifi.html#_CPPv414wifi_country_t)|
|`esp_wifi_get_max_tx_power()` = 78 by default at initial power on|0.25 dBm|[esp_wifi_get_max_tx_power](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_wifi.html#_CPPv425esp_wifi_get_max_tx_powerP6int8_t)|
|`WiFi.getTxPower()`|0.25 dBm|Based on value extracted|
|`WiFi.setTxPower()`|0.25 dBm|Based on `getTxPower()`|

So to answer my own question:
``
if the country should be limited to 30, then why is the board doing 78 at runtime?
``
The answer is: min_value(30 * 4, 78) = 78.
That is why. The board is limited at 78. 
78 * 0.25 = 19.5 dBm.
... or something along those lines.

**Reply by frankonski** [reply]
So I finally tested the power with a volt meter. I'm getting contant 13.67V from the 12V panel, and getting 4.97 from the 5v regulator. I tested for about 30 seconds, didn't see any fluctuation. But, when reconnecting the wrires to the breadboard I kinda felt more resistance when inserting the wires back in different holes. I wonder if the brownouts were caused by weak connections from the breadboard. 

I will be monitoring it for a while. If it last 2-3 days this way without any brownouts, I'd say that was the cause from the get-go. :(

**Reply by frankonski** [reply]
well, one more router restart this morning, one more brownout while not connected with USB adapter. Is there a way to save all weblogs to nvram (when logged) so that on restart, previous logs remain available? I can't just connect the dev board to USB and turn on serial logging: the board will get power from USB and will not brownout.

**Reply by HomeSpan** [reply]
There is currently no easy way to save WebLog in NVS (would need substantive change to HomeSpan).  Do you have capacitors on your 7805 and a heat sink?  (see for example, https://www.adafruit.com/product/2164, as well page 4 of https://www.mouser.com/datasheet/2/389/l78-1849632.pdf).

**Reply by frankonski** [reply]
No capacitor, no extra heat sink. Purchasing a new meter to see amps.

**Reply by HomeSpan** [reply]
You probably won't be able to see transient spikes that causes a brownout with a meter - would need a digital scope with a trigger and saving mechanism.

**Reply by frankonski** [reply]
Something odd is happening with the regulator. Nevertheless, I think I am now confident the WiFi TX power has nothing to do with brownouts I have been seeing. I'm ok to close that discussion.

**Reply by HomeSpan** [reply]
Thanks for update.  Let us know if you solve the regulator issue.

**Reply by frankonski** [reply, marked as answer]
ok, well, after connecting the ESP32 to the 12v backup battery of the alarm panel, rather than connecting to the 12v aux(+) of that same alarm panel, no more brownouts. Btw, I had also added capacitors while connected to the panel directly and was still seeing brownouts. So regulator & capacitor was not the issue. The output coming from the panel is / was the problem.

And so far, connected to the backup battery, with regulator & capacitors, everything seems stable, even on wifi disconnects / reconnects.

Case closed.

----
Update
-----
3-weeks later update: a heat sink on the regulator is required. Even with surrounding capacitors the brownouts were still happening. After more troubleshooting, and being told by someone with electrical engineering skills, it looks like 60C for the regulator was too high (too hot to the touch). I installed some home-made heatsink on it and that automatically resolved the occuring brownouts. Haven't had one in 3 weeks despites many wifi reconnects and re-flashing.

**Comment by frankonski** [comment]
Wait, something strange. Since I posted that initial question above, went for dinner, and hit refresh, a new brownout occured few seconds ago.
But now I see this logged:
`Country code: CN, country maxTxPower: 20 (ESP 32 maxTxPower: 52, currentTxPower: 52), policy: AUTO`
..followed by 
`Country code: US, country maxTxPower: 30 (ESP 32 maxTxPower: 52, currentTxPower: 52), policy: AUTO`

I will have to reproduce, but I wonder if `78` is right after flashing, and `52` is after.. brownout??

**Reply by frankonski** [reply]
And to add to the wifi oddity. Watching a movie while waiting for your much-appreciated-and-valuable responses.
Turn the laptop lid on, hit refresh on the HomeSpan WebLogs page (/status) and I see a brownout 10 seconds ago.

I know (or think) Chrome will re-establish TCP connections earlier, maybe before hitting Refresh, just to be faster in returning a response. The timing matches. Like if waking up that old crappy laptop would trigger an HTTP re-connection to the weblogs, causing that brownout. This is really strange and coïncidental.
